crates/biome_js_analyze/src/lint/nursery/use_import_extensions.rs

```diff
@@ -0,0 +1,201 @@
+use std::path::Path;
+
+use biome_analyze::{
+    context::RuleContext, declare_rule, ActionCategory, Ast, Rule, RuleDiagnostic,
+};
+use biome_console::markup;
+use biome_diagnostics::Applicability;
+use biome_js_factory::make;
+use biome_js_syntax::JsModuleSource;
+use biome_rowan::{AstNode, BatchMutationExt, TokenText};
+
+use crate::JsRuleAction;
+
+use super::use_import_restrictions::get_basename;
+
+declare_rule! {
+    /// Require import extensions for relative imports.
+    ///
+    /// Browsers and Node.js do not natively support importing files without extensions. This rule
+    /// enforces the use of import extensions for relative imports to make the code more consistent.
+    ///
+    /// Tooling also benefits from explicit import extensions, because they do not need to guess which
+    /// file to resolve.
+    ///
+    /// ## Examples
+    ///
+    /// ### Invalid
+    ///
+    /// ```js,expect_diagnostic
+    /// import "./foo";
+    /// ```
+    /// ```js,expect_diagnostic
+    /// import "./bar/";
+    /// ```
+    /// ```js,expect_diagnostic
+    /// import "../";
+    /// ```
+    /// ```js,expect_diagnostic
+    /// import "../.";
+    /// ```
+    ///
+    /// ### Valid
+    ///
+    /// ```js
+    /// import "biome";
+    /// ```
+    /// ```js
+    /// import "./foo.js";
+    /// ```
+    /// ```js
+    /// import "./bar/index.js";
+    /// ```
+    ///
+    /// ## Caveats
+    ///
+    /// Rule does not yet check filesystem for file type. It tries to guess which extension
+    /// it should add based on the file extension of the current file and the import path.
+    /// When applying the suggested fix, make sure to verify that the file type is correct.
+    ///
+    pub UseImportExtensions {
+        version: "next",
+        name: "useImportExtensions",
+        recommended: false,
+    }
+}
+
+impl Rule for UseImportExtensions {
+    type Query = Ast<JsModuleSource>;
+    type State = UseImportExtensionsState;
+    type Signals = Option<Self::State>;
+    type Options = ();
+
+    fn run(ctx: &RuleContext<Self>) -> Self::Signals {
+        let binding = ctx.query();
+
+        let Ok(module_path) = binding.inner_string_text() else {
+            return None;
+        };
+
+        let file_ext = ctx.file_path().extension()?.to_str()?;
+
+        get_extensionless_import(file_ext, &module_path)
+    }
+
+    fn diagnostic(ctx: &RuleContext<Self>, _: &Self::State) -> Option<RuleDiagnostic> {
+        let node = ctx.query();
+
+        Some(
+            RuleDiagnostic::new(
+                rule_category!(),
+                node.range(),
+                markup! {
+                    "Import extension is required for relative imports."
```

Comment:
We could avoid the passive voice here:
```suggestion
                    "Add a file extension for relative imports."
```